### PR TITLE
Implement LogitProcessor, Expose ForwardTokensAndSample, and support customRequest in web worker

### PR DESCRIPTION
**Pull request by rickzx**
This PR 
- introduces LogitProcessor interface:
```
interface LogitProcessor {
  processLogits: (logits: Float32Array) => Float32Array;
  processSampledToken: (token: number) => void;
  resetState: () => void;
}
```
which can be passed to `ChatModule` to allow pre-processing of the logits before sampling tokens. 

Note that the LogitProcessor can be stateful, meaning that the processing rule can depend on things like past tokens sampled, etc.

Besides, we:

- Expose a lower-level API `forwardTokensAndSample()` to generate tokens and sample them.
- Support an additional `kind` in web worker messages `customRequest` to allow customized communications between the main thread and the worker thread apart from the predefined ones.

We also add an example of how to use a `LogitProcessor`, both with a web worker and without.

Co-authored-by: @CharlieFRuan

### CODE REVIEW

**Review by CharlieFRuan — APPROVED** [review]
Thanks! Let's merge and update the npm.

**Comment by CharlieFRuan** [comment]
Note to self: `LogitProcessor` is only mentioned in `src/chat_module.ts` and not a part of `src/types.ts`.
